### PR TITLE
alpha-1

### PR DESCRIPTION
**Pull request by Swap76**
**Adding the following modules**

- Dependencies and go.mod setup
- Configuration (config.go)
- Logger (logger.go)
- Health Handler (health_handler.go)
- Version Handler (version_handler.go)

**Need review about** 

- The current structuring  
- Should we make the repo shallow (remove src and bring everything out) as this will allow us to use simple Heroku building process
- If not removing src then which method to use makefile, vendor, etc

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1 @@
+PORT=
```
Comment (prabhuomkar) [inline .env.example]: Give a dummy value to make it easy to run on localhost.

File: Procfile [hunk 2]
```diff
@@ -0,0 +1 @@
+web: epicentre
```
Comment (prabhuomkar) [inline Procfile]: does this work? as it expects the binary path like `/bin/*` I believe.

File: src/config/config.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"fmt"
+	"github.com/gin-gonic/gin"
+	env "github.com/caarlos0/env/v6"
+)
+
+// Config Available everywhere
+var Config *MainConfig
+
+type (
```
Comment (prabhuomkar) [inline src/config/config.go]: For single type you can write `type XYZ struct {`

File: src/config/config.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	// MainConfig Type exported for use as input type in functions
+	MainConfig struct {
+		Port       string			`env:"PORT" envDefault:"8000"`
+		FileName   string			`env:"FILENAME" envDefault:"zap.log"`
```
Comment (prabhuomkar) [inline src/config/config.go]: envDefault should be `epicentre.log`

File: src/config/config.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		MaxBackUp  int				`env:"MAZBACKUP" envDefault:"5"`
+		Compress   bool				`env:"COMPRESS" envDefault:"false"`
+		Level      string			`env:"LEVEL" envDefault:"info"`
+		OutputType string			`env:"OUTPUTTYPE" envDefault:"json"`
+	}
+)
+
+// LoadConfig Loads the config
+func LoadConfig(router *gin.Engine)  {
+	fmt.Println("Initializing LoadConfig")
```
Comment (prabhuomkar) [inline src/config/config.go]: `Initializing configuration variables...`

File: src/config/config.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	cfg := MainConfig{}
+	if err := env.Parse(&cfg); err != nil {
+		panic(err)
+	}
+
+	// Address of parsed configs passing to Global Config variable
+	Config = &cfg
+
+	// Configuring Static Assets for development
+	router.Static("/assets", "../images")
```
Comment (prabhuomkar) [inline src/config/config.go]: I believe we don't need favicon and assets for our APIs, will these be removed in future?

File: src/main.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/siesgstarena/epicentre/src/config"
+	routes "github.com/siesgstarena/epicentre/src/router"
+	"github.com/siesgstarena/epicentre/src/services/logger"
+)
+
+func main() {
+
+	router := gin.Default()
```
Comment (prabhuomkar) [inline src/main.go]: Order should be:
1. Config
2. Logger
3. Kafka
4. MongoDB
5. Router

File: src/config/config.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		MaxAge     int				`env:"MAXAGE" envDefault:"10"`
+		MaxBackUp  int				`env:"MAZBACKUP" envDefault:"5"`
+		Compress   bool				`env:"COMPRESS" envDefault:"false"`
+		Level      string			`env:"LEVEL" envDefault:"info"`
+		OutputType string			`env:"OUTPUTTYPE" envDefault:"json"`
+	}
+)
+
+// LoadConfig Loads the config
+func LoadConfig(router *gin.Engine)  {
```
Comment (prabhuomkar) [inline src/config/config.go]: return value should be `error`

File: src/config/config.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	fmt.Println("Initializing LoadConfig")
+
+	cfg := MainConfig{}
+	if err := env.Parse(&cfg); err != nil {
+		panic(err)
```
Comment (prabhuomkar) [inline src/config/config.go]: return this error and don't panic here.

File: src/web/health_handler.go [hunk 10]
```diff
@@ -0,0 +1,26 @@
+package web
+
+import (
+	"time"
+	"github.com/gin-gonic/gin"
+	"github.com/siesgstarena/epicentre/src/services/logger"
+)
+
+type health struct {
+    Name     		string `json:"name"`
```
Comment (prabhuomkar) [inline src/web/health_handler.go]: run `go fmt ./...` inside `src`

File: src/main.go [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+	config.LoadConfig(router)
```
Comment (prabhuomkar) [inline src/main.go]: use the return value from config to panic if err here.

File: src/web/health_handler.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+	Health 			string `json:"health"`
+	Timestamp       time.Time `json:"timestamp"`
+}
+
+// HeathHandler Sends info about health of API
+func HeathHandler(c *gin.Context)  {
+	logger.Log.Info("Inside Health Handler")
+	info := new(health)
+	info.Name = "epicentre"
+	info.Description = "Cloud Monitoring and Alerting Tool built by SIESGSTarena Platform Team)"
```
Comment (prabhuomkar) [inline src/web/health_handler.go]: why `)` in the end?

File: src/web/health_handler.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+type health struct {
+    Name     		string `json:"name"`
+	Description     string `json:"description"`
+	Health 			string `json:"health"`
+	Timestamp       time.Time `json:"timestamp"`
+}
+
+// HeathHandler Sends info about health of API
+func HeathHandler(c *gin.Context)  {
+	logger.Log.Info("Inside Health Handler")
```
Comment (prabhuomkar) [inline src/web/health_handler.go]: don't have to log inside handlers unless custom or complex operations.. we will use gin to log http requests.

File: src/web/version_handler.go [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+// VersionHandler Sends info about version of API
+func VersionHandler(c *gin.Context)  {
+	logger.Log.Info("Inside Health Handler")
+	info := new(version)
+	info.Name = "epicentre"
+	info.Description = "Cloud Monitoring and Alerting Tool built by SIESGSTarena Platform Team)"
+	info.Version = runtime.Version()
+	loc, _ := time.LoadLocation("UTC")
+    info.Timestamp = time.Now().In(loc)
```
Comment (prabhuomkar) [inline src/web/version_handler.go]: `go fmt` here as well :)

**Review by prabhuomkar — CHANGES_REQUESTED** [review]

File: Procfile (hunk 2 above)
Comment (Swap76) [reply Procfile]: Actually I tried several version in between commits the correct should be `/bin/*` reverting to that

File: src/config/config.go (hunk 6 above)
Comment (Swap76) [reply src/config/config.go]: Sure, The error when loading on the browser was getting as it was checking for the favicon. But got that when consuming as API we don't need this :-)

File: src/web/health_handler.go (hunk 13 above)
Comment (Swap76) [reply src/web/health_handler.go]: Sure. Gin is logging our every request just like morgan

File: src/web/version_handler.go (hunk 14 above)
Comment (Swap76) [reply src/web/version_handler.go]: In my laptop their some issue of fmt I guess it doesn't always modify files.

But this indentation error is not seen in my laptop don't know-how

File: src/web/health_handler.go (hunk 10 above)
Comment (Swap76) [reply src/web/health_handler.go]: This Indentation error of health also is not visible in my machine

**Review by prabhuomkar — APPROVED** [review]

**Comment by Swap76** [comment]
Deployment Changed to Master Branch as Auto Deploy

**Pull request merged**